File: bend2/effs/tcp_listen.js
// TCP
// ===

// TCP.listen(port) binds every interface, TCP.listen_on(host, port) the
// one dotted IPv4 address named; the shared pair sets SO_REUSEPORT too.
function tcp_listen_at(host, port, shared) {
  const sys = io_sys();
  const fd = sys.socket(2, 1, 0);
  if (fd < 0) {
    return io_fail(sys.errno());
  }
  const one = new Int32Array([1]);
  const level = sys.mac ? 0xffff : 1;
  sys.setsockopt(fd, level, sys.mac ? 4 : 2, sys.ptr(one), 4);
  if (shared) {
    sys.setsockopt(fd, level, sys.mac ? 0x200 : 15, sys.ptr(one), 4);
  }
  const at = io_addr(host, Number(port));
  if (at === null) {
    sys.close(fd);
    return io_fail(22);
  }
  // the backlog is clamped to the kernel's limit (somaxconn)
  if (sys.bind(fd, sys.ptr(at), 16) < 0 || sys.listen(fd, 4096) < 0
    || sys.fcntl(fd, 4, sys.fcntl(fd, 3, 0) | (sys.mac ? 4 : 0x800)) < 0) {
    const code = sys.errno();
    sys.close(fd);
    return io_fail(code);
  }
  return io_done(fd);
}

function tcp_listen(port) {
  return tcp_listen_at("0.0.0.0", port, false);
}

function tcp_listen_on(host, port) {
  return tcp_listen_at(host, port, false);
}

function tcp_listen_shared(port) {
  return tcp_listen_at("0.0.0.0", port, true);
}

function tcp_listen_shared_on(host, port) {
  return tcp_listen_at(host, port, true);
}

File: bend2/effs/tcp_listen.c
// TCP
// ===

// TCP.listen(port) binds every interface; TCP.listen_on(host, port) the
// one dotted IPv4 address named (EINVAL for anything else).
uint32_t tcp_listen_at(const char* host, uint32_t port, int* out) {
  int fd = socket(AF_INET, SOCK_STREAM, 0);
  if (fd < 0) {
    return (uint32_t)errno;
  }
  int one = 1;
  setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one));
  struct sockaddr_in at;
  if (io_sys_addr(host, port, &at) < 0) {
    close(fd);
    return EINVAL;
  }
  int bound = bind(fd, (struct sockaddr*)&at, sizeof(at));
  if (bound < 0 || listen(fd, SOMAXCONN) < 0
    || fcntl(fd, F_SETFL, fcntl(fd, F_GETFL) | O_NONBLOCK) < 0) {
    uint32_t code = (uint32_t)errno;
    close(fd);
    return code;
  }
  *out = fd;
  return 0;
}

uint32_t tcp_listen(uint32_t port, int* out) {
  return tcp_listen_at("0.0.0.0", port, out);
}

static Term tcp_listen_end(Env e, uint32_t q, int out) {
  return q != 0 ? io_fail(e, q, NULL) : io_done(e, io_hand(out));
}

#ifdef CID_TCP_LISTEN
Term tcp_listen_run(Env e, Term* f, IoWork* w) {
  int out = -1;
  uint32_t q = tcp_listen((uint32_t)f[0], &out);
  return tcp_listen_end(e, q, out);
}

static void __attribute__((constructor)) tcp_listen_use(void) {
  io_eff(CID_TCP_LISTEN, tcp_listen_run, 0);
}
#endif

#ifdef CID_TCP_LISTEN_ON
Term tcp_listen_on_run(Env e, Term* f, IoWork* w) {
  u64   n = 0;
  char* host = io_cstr(e, f[0], &n);
  int   out = -1;
  uint32_t q = io_nul(host, n) ? EINVAL : tcp_listen_at(host, (uint32_t)f[1], &out);
  free(host);
  return tcp_listen_end(e, q, out);
}

static void __attribute__((constructor)) tcp_listen_on_use(void) {
  io_eff(CID_TCP_LISTEN_ON, tcp_listen_on_run, 0);
}
#endif
